{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2b6c3d44-aaef-4f33-9997-be19b3b1203a",
   "metadata": {},
   "source": [
    "# Processing Whole Exome Sequencing (WES/WXS) Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f295e34f-eb78-4509-9d4a-ccfbb8e25146",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "     active environment : sra-tools\n",
      "    active env location : /home/fmbuga/.conda/envs/sra-tools\n",
      "            shell level : 3\n",
      "       user config file : /home/fmbuga/.condarc\n",
      " populated config files : /home/fmbuga/.condarc\n",
      "          conda version : 4.10.3\n",
      "    conda-build version : not installed\n",
      "         python version : 3.6.9.final.0\n",
      "       virtual packages : __linux=3.10.0=0\n",
      "                          __glibc=2.17=0\n",
      "                          __unix=0=0\n",
      "                          __archspec=1=x86_64\n",
      "       base environment : /opt/intel/intelpython3  (read only)\n",
      "      conda av data dir : /opt/intel/intelpython3/etc/conda\n",
      "  conda av metadata url : None\n",
      "           channel URLs : https://conda.anaconda.org/conda-forge/linux-64\n",
      "                          https://conda.anaconda.org/conda-forge/noarch\n",
      "                          https://conda.anaconda.org/bioconda/linux-64\n",
      "                          https://conda.anaconda.org/bioconda/noarch\n",
      "                          https://repo.anaconda.com/pkgs/main/linux-64\n",
      "                          https://repo.anaconda.com/pkgs/main/noarch\n",
      "                          https://repo.anaconda.com/pkgs/r/linux-64\n",
      "                          https://repo.anaconda.com/pkgs/r/noarch\n",
      "                          https://conda.anaconda.org/main/linux-64\n",
      "                          https://conda.anaconda.org/main/noarch\n",
      "          package cache : /opt/intel/intelpython3/pkgs\n",
      "                          /home/fmbuga/.conda/pkgs\n",
      "       envs directories : /home/fmbuga/.conda/envs\n",
      "                          /opt/intel/intelpython3/envs\n",
      "               platform : linux-64\n",
      "             user-agent : conda/4.10.3 requests/2.25.1 CPython/3.6.9 Linux/3.10.0-1062.18.1.el7.x86_64 centos/7.9.2009 glibc/2.17\n",
      "                UID:GID : 1179:1090\n",
      "             netrc file : None\n",
      "           offline mode : False\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1f9a893-eea7-4565-9e9b-34037451ac0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/fmbuga/ERR1831349\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5ebefd6a-11f3-4fb9-83dd-9bb84b0f4f60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "230722_P_fastq_2_vcf.ipynb\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "938dbcd4-71f8-4000-9944-bba5b344980e",
   "metadata": {},
   "source": [
    "## Download data as fastq from NCBI SRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3b93270b-abc7-434f-bfa2-45534d14476b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=fasterq_dump_230722\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/fasterq_dump_230722_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/fasterq_dump_230722_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=28\n",
      "#SBATCH --mem=0\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate sra-tools                      ### CONDA ENVIRONMENT NAME\n",
      "echo \"\"\n",
      "\n",
      "work_dir=$(pwd)\n",
      "echo \"working directory: $work_dir\"     ### WORKING DIRECTORY\n",
      "echo \"\"\n",
      "\n",
      "ncbi_sra_accession=$1                   ### NCBI SRA archive accession ID to download\n",
      "\n",
      "# fasterq-dump\n",
      "call=$(fasterq-dump $ncbi_sra_accession \\\n",
      "    --threads 28 \\\n",
      "    --progress \\\n",
      "    --details \\\n",
      "    --split-files)                                   ### split-files for paired-end data\n",
      "    \n",
      "echo $call\n",
      "eval $call\n",
      "echo \"fasterq-dump COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "\n",
      "conda info\n",
      "echo \"\"\n",
      "conda list\n",
      "echo \"\"\n",
      "\n",
      "##################\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat fasterq_dump_230722_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9325fa3a-bf6c-4a2b-9fe1-5228086f486e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 645837\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645836     nodes     bash   fmbuga  R    1:08:45      1 node18\n",
      "            645837     nodes fasterq_   fmbuga  R       0:05      1 node35\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645836     nodes     bash   fmbuga  R    1:08:55      1 node18\n",
      "            645837     nodes fasterq_   fmbuga  R       0:15      1 node35\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645836     nodes     bash   fmbuga  R    1:09:11      1 node18\n",
      "            645837     nodes fasterq_   fmbuga  R       0:31      1 node35\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645836     nodes     bash   fmbuga  R    1:09:41      1 node18\n",
      "            645837     nodes fasterq_   fmbuga  R       1:01      1 node35\n"
     ]
    }
   ],
   "source": [
    "# submit fasterq-dump job to SLURM\n",
    "\n",
    "!sbatch fasterq_dump_230722_slurm.sh \\\n",
    "    ERR1831349\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "885e743e-4ac1-490a-a21f-d154782926b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR1831349_1.fastq  ERR1831349_2.fastq\n"
     ]
    }
   ],
   "source": [
    "# create folder to store raw fastq and move downloaded files there\n",
    "!mkdir ./00_raw_fastq\n",
    "!mv *.fastq ./00_raw_fastq\n",
    "!ls ./00_raw_fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df8e2be4-f5c5-4c56-ac58-0cb6502abf1b",
   "metadata": {},
   "source": [
    "## Check Quality of fastq with FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "998d85ad-ad11-4f41-bc22-ef3be70ab2e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=fastqc_allInFolder_230723\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/fastqc_allInFolder_230723_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/fastqc_allInFolder_230723_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=2\n",
      "#SBATCH --mem=0\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate sra-tools                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "call=\"fastqc --version\"\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "input_dir=$1             ### INPUT DIRECTORY\n",
      "output_dir=$2            ### OUTPUT DIRECTORY \n",
      "echo \"output directory: $output_dir\"\n",
      "echo \"\"\n",
      "\n",
      "# fastqc\n",
      "\n",
      "call=$(find $input_dir -name \"*.fastq\" | \\\n",
      "        xargs fastqc \\\n",
      "                --outdir $output_dir \\\n",
      "                --threads 2)                                   ### INPUT FASTQs\n",
      "    \n",
      "echo $call\n",
      "eval $call\n",
      "echo \"fastqc COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat fastqc_allinFolder_230723_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6b8b75fa-4cc4-49b5-9ffd-15d53c59e571",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 645885\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645884     nodes     bash   fmbuga  R      33:05      1 node59\n",
      "            645885     nodes fastqc_a   fmbuga  R       0:05      1 node60\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645884     nodes     bash   fmbuga  R      33:15      1 node59\n",
      "            645885     nodes fastqc_a   fmbuga  R       0:15      1 node60\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645884     nodes     bash   fmbuga  R      33:30      1 node59\n",
      "            645885     nodes fastqc_a   fmbuga  R       0:30      1 node60\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645884     nodes     bash   fmbuga  R      34:01      1 node59\n",
      "            645885     nodes fastqc_a   fmbuga  R       1:01      1 node60\n"
     ]
    }
   ],
   "source": [
    "# submit fastqc job to SLURM\n",
    "!sbatch fastqc_allinFolder_230723_slurm.sh \\\n",
    "    ./00_raw_fastq \\\n",
    "    ./00_raw_fastq\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "670a65ec-7f61-46f8-8084-9ec91c038792",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries for HTML visualization\n",
    "from IPython.display import IFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "07d24c40-e0b8-4983-ba45-9d7c3de64b8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"1500\"\n",
       "            height=\"2000\"\n",
       "            src=\"./00_raw_fastq/ERR1831349_1_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x2b935913e490>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IFrame(src='./00_raw_fastq/ERR1831349_1_fastqc.html', width=1500, height=2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6b7d7840-ac53-4ea6-ba13-ddbb4962b410",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"1500\"\n",
       "            height=\"2000\"\n",
       "            src=\"./00_raw_fastq/ERR1831349_2_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x2b9357cf8290>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IFrame(src='./00_raw_fastq/ERR1831349_2_fastqc.html', width=1500, height=2000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60a49461-e655-4696-9d0b-ee298b595660",
   "metadata": {},
   "source": [
    "- since no red flags in FastQC, proceed with mapping\n",
    "- if mapping doesn't pass QC checks may return and see if some trimming helps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "426fa18d-9338-45e9-a562-4a42560acec2",
   "metadata": {},
   "source": [
    "## map reads to human reference hg38 with bwa mem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fea9dece-26df-474a-938c-155b486af7b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=bwa_mem_hg38_samtools_fastq_to_bam_230723\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/bwa_mem_hg38_samtools_fastq_to_bam_230723_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/bwa_mem_hg38_samtools_fastq_to_bam_230723_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=28\n",
      "##SBATCH --mem-per-cpu=4G\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate gatk4                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "call=\"bwa\" # to get the version\n",
      "echo \"bwa version\"\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "call=\"samtools --version\"\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "fastq1=$1                             ### INPUT FASTQ1\n",
      "fastq2=$2                             ### INPUT FASTQ2\n",
      "output_dir=$3                         ### OUTPUT DIRECTORY \n",
      "\n",
      "echo \"output directory: $output_dir\"\n",
      "echo \"\"\n",
      "\n",
      "out_name=$(basename $fastq1 _1.fastq)\n",
      "echo $out_name\n",
      "\n",
      "echo $output_dir$out_name.bam\n",
      "\n",
      "# bwa mem --> samtools\n",
      "\n",
      "call=$(bwa mem \\\n",
      "    -t 8 \\\n",
      "    -R \"@RG\\tID:$out_name\\tSM:$out_name\" \\\n",
      "    ~/tools/hg38/hg38.fa \\\n",
      "    $fastq1 $fastq2 | \\\n",
      "        samtools view -uh | \\\n",
      "            samtools sort \\\n",
      "                -o \"$output_dir$out_name\".bam \\\n",
      "                -@ 28 \\\n",
      "                --write-index \\\n",
      "                --verbosity 9 \\\n",
      "                -m 2G)                                   ### INPUT FASTQs\n",
      "    \n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"mapping to ref, and sorting and indexing bam COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat bwa_mem_hg38_samtools_fastq_to_bam_230723_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "203d30de-2267-43db-a239-d97a90ed04e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create directory to output bam files to\n",
    "!mkdir ./01_raw_bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0ae7a7e1-7ea3-454b-88ea-87561ace50ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 645893\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R      28:21      1 node58\n",
      "            645893     nodes bwa_mem_   fmbuga  R       0:05      1 node05\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R      28:31      1 node58\n",
      "            645893     nodes bwa_mem_   fmbuga  R       0:15      1 node05\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R      28:46      1 node58\n",
      "            645893     nodes bwa_mem_   fmbuga  R       0:30      1 node05\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R      29:16      1 node58\n",
      "            645893     nodes bwa_mem_   fmbuga  R       1:00      1 node05\n"
     ]
    }
   ],
   "source": [
    "# submit mapping job to SLURM HPC\n",
    "!sbatch bwa_mem_hg38_samtools_fastq_to_bam_230723_slurm.sh \\\n",
    "    ./00_raw_fastq/ERR1831349_1.fastq \\\n",
    "    ./00_raw_fastq/ERR1831349_2.fastq \\\n",
    "    ./01_raw_bam/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87122898-4a7c-4e77-afad-1995ff65fc87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93603695 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "93556700 + 0 primary\n",
      "0 + 0 secondary\n",
      "46995 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "93403023 + 0 mapped (99.79% : N/A)\n",
      "93356028 + 0 primary mapped (99.79% : N/A)\n",
      "93556700 + 0 paired in sequencing\n",
      "46778350 + 0 read1\n",
      "46778350 + 0 read2\n",
      "92600884 + 0 properly paired (98.98% : N/A)\n",
      "93279060 + 0 with itself and mate mapped\n",
      "76968 + 0 singletons (0.08% : N/A)\n",
      "593850 + 0 with mate mapped to a different chr\n",
      "517790 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "# QC check output bam file\n",
    "!samtools flagstat \\\n",
    "    -@ 28 \\\n",
    "    ./01_raw_bam/ERR1831349.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0583f50-39d0-492f-9973-e5c9558499dd",
   "metadata": {},
   "source": [
    "- based on the high mapping rate (~99%) no need to go back and trim fastq reads\n",
    "- proceed with marking duplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39536291-1368-4ce1-9d22-9c298de765c8",
   "metadata": {},
   "source": [
    "## Mark Duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8494cd88-224f-48cf-a20a-f4ae0ae2cb57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=samtools_collate_fixmate_sort_markdup_221222\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_collate_fixmate_sort_markdup_221222_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_collate_fixmate_sort_markdup_221222_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=28\n",
      "#SBATCH --mem-per-cpu=4G\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate minimap2_224                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "call=\"samtools --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "input_bam=$1\n",
      "output_dir=$2\n",
      "\n",
      "out_name=$(basename $input_bam .bam)\n",
      "\n",
      "call=$(samtools collate \\\n",
      "    -Ou \\\n",
      "    --threads 28 \\\n",
      "    $input_bam | \\\n",
      "        samtools fixmate \\\n",
      "            - \\\n",
      "            - \\\n",
      "            --threads 28 \\\n",
      "            -u | \\\n",
      "            samtools sort \\\n",
      "                - \\\n",
      "                --threads 28 \\\n",
      "                -u |\\\n",
      "                samtools markdup \\\n",
      "                    - \\\n",
      "                    --threads 28 \\\n",
      "                    -sf \"$output_dir$out_name\"_collate_fixmate_sort_markdup_bam.markdupstats \\\n",
      "                    --write-index \\\n",
      "                    \"$output_dir$out_name\"_collate_fixmate_sort_markdup.bam)\n",
      "\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"samtools collate > fixmate > sort > markdup COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat ~/tools/slurm_scripts/samtools_collate_fixmate_sort_markdup_221222_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1925a6d6-9fe9-4e6e-9049-980f4ea20bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy mark duplicates shell script\n",
    "!cp ~/tools/slurm_scripts/samtools_collate_fixmate_sort_markdup_221222_slurm.sh ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de867a03-a48c-4d6f-8db7-00d3cb154835",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make folder to store bam with duplicates marked\n",
    "!mkdir ./02_bam_markdup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "93b8338b-0b21-4b63-a97a-272fa16364b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 645897\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    1:44:02      1 node58\n",
      "            645897     nodes samtools   fmbuga  R       0:05      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    1:44:12      1 node58\n",
      "            645897     nodes samtools   fmbuga  R       0:15      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    1:44:27      1 node58\n",
      "            645897     nodes samtools   fmbuga  R       0:30      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    1:44:57      1 node58\n",
      "            645897     nodes samtools   fmbuga  R       1:00      1 node62\n"
     ]
    }
   ],
   "source": [
    "# submit mark duplicates job to SLURM HPC\n",
    "!sbatch samtools_collate_fixmate_sort_markdup_221222_slurm.sh \\\n",
    "    ./01_raw_bam/ERR1831349.bam \\\n",
    "    ./02_bam_markdup/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "879f2ab3-7371-487f-b2d3-eac8bb4eeb2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 645898\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    1:54:17      1 node58\n",
      "            645898     nodes samtools   fmbuga  R       0:05      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    1:54:27      1 node58\n",
      "            645898     nodes samtools   fmbuga  R       0:15      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    1:54:43      1 node58\n",
      "            645898     nodes samtools   fmbuga  R       0:31      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    1:55:13      1 node58\n",
      "            645898     nodes samtools   fmbuga  R       1:01      1 node62\n"
     ]
    }
   ],
   "source": [
    "# submit mark duplicates job to SLURM HPC\n",
    "!sbatch samtools_collate_fixmate_sort_markdup_221222_slurm.sh \\\n",
    "    ~/nanopore/rogstrix/dorado/01_bam_raw/FAF01132-84868110.bam \\\n",
    "    ./02_bam_markdup/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ff2c1e8c-6247-4634-ac02-77c748f14fa7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1681556 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "719983 + 0 primary\n",
      "505430 + 0 secondary\n",
      "456143 + 0 supplementary\n",
      "17979 + 0 duplicates\n",
      "17979 + 0 primary duplicates\n",
      "1617881 + 0 mapped (96.21% : N/A)\n",
      "656308 + 0 primary mapped (91.16% : N/A)\n",
      "0 + 0 paired in sequencing\n",
      "0 + 0 read1\n",
      "0 + 0 read2\n",
      "0 + 0 properly paired (N/A : N/A)\n",
      "0 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (N/A : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "# QC check output bam file\n",
    "!samtools flagstat \\\n",
    "    -@ 28 \\\n",
    "    ./02_bam_markdup/FAF01132-84868110_collate_fixmate_sort_markdup.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0c897adf-becf-4969-be4b-9cb76166b811",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1681556 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "719983 + 0 primary\n",
      "505430 + 0 secondary\n",
      "456143 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "1617881 + 0 mapped (96.21% : N/A)\n",
      "656308 + 0 primary mapped (91.16% : N/A)\n",
      "0 + 0 paired in sequencing\n",
      "0 + 0 read1\n",
      "0 + 0 read2\n",
      "0 + 0 properly paired (N/A : N/A)\n",
      "0 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (N/A : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "# QC check output bam file\n",
    "!samtools flagstat \\\n",
    "    -@ 28 \\\n",
    "    ~/nanopore/rogstrix/dorado/01_bam_raw/FAF01132-84868110.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8cab7e70-7328-48e9-8ded-695ba92af44f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=samtools_collate_fixmate_sort_markdup_230723\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_collate_fixmate_sort_markdup_230723_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_collate_fixmate_sort_markdup_230723_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=28\n",
      "#SBATCH --mem=0\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate minimap2_224                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "call=\"samtools --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "input_bam=$1\n",
      "output_dir=$2\n",
      "\n",
      "out_name=$(basename $input_bam .bam)\n",
      "\n",
      "call=$(samtools collate \\\n",
      "    -Ou \\\n",
      "    --threads 28 \\\n",
      "    $input_bam | \\\n",
      "        samtools fixmate \\\n",
      "            - \\\n",
      "            - \\\n",
      "            --threads 28 \\\n",
      "            -u | \\\n",
      "            samtools sort \\\n",
      "                - \\\n",
      "                --threads 28 \\\n",
      "                -u |\\\n",
      "                samtools markdup \\\n",
      "                    - \\\n",
      "                    --threads 28 \\\n",
      "                    -sf \"$output_dir$out_name\"_collate_fixmate_sort_markdup_bam.markdupstats \\\n",
      "                    --write-index \\\n",
      "                    \"$output_dir$out_name\"_collate_fixmate_sort_markdup.bam)\n",
      "\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"samtools collate > fixmate > sort > markdup COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat samtools_collate_fixmate_sort_markdup_230723_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "56ac1449-4801-4f8e-8c32-a3e3b9991cbe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 645899\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    2:26:44      1 node58\n",
      "            645899     nodes samtools   fmbuga  R       0:05      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    2:26:54      1 node58\n",
      "            645899     nodes samtools   fmbuga  R       0:15      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    2:27:09      1 node58\n",
      "            645899     nodes samtools   fmbuga  R       0:30      1 node62\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            645891     nodes     bash   fmbuga  R    2:27:39      1 node58\n",
      "            645899     nodes samtools   fmbuga  R       1:00      1 node62\n"
     ]
    }
   ],
   "source": [
    "# submit mark duplicates job to SLURM HPC\n",
    "!sbatch samtools_collate_fixmate_sort_markdup_221222_slurm.sh \\\n",
    "    ./01_raw_bam/ERR1831349.bam \\\n",
    "    ./02_bam_markdup/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ef7e21ca-7742-413a-86b2-a3e1b68fedba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=samtools_collate_fixmate_sort_markdup_HIMEM_230724\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_collate_fixmate_sort_markdup_HIMEM_230724_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_collate_fixmate_sort_markdup_HIMEM_230724_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=12\n",
      "#SBATCH --mem=0\n",
      "#SBATCH --partition=himem\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate minimap2_224                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "call=\"samtools --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "input_bam=$1\n",
      "output_dir=$2\n",
      "\n",
      "out_name=$(basename $input_bam .bam)\n",
      "\n",
      "call=$(samtools collate \\\n",
      "    -Ou \\\n",
      "    --threads 12 \\\n",
      "    $input_bam | \\\n",
      "        samtools fixmate \\\n",
      "            - \\\n",
      "            - \\\n",
      "            --threads 12 \\\n",
      "            -u | \\\n",
      "            samtools sort \\\n",
      "                - \\\n",
      "                --threads 12 \\\n",
      "                -u |\\\n",
      "                samtools markdup \\\n",
      "                    - \\\n",
      "                    --threads 12 \\\n",
      "                    -sf \"$output_dir$out_name\"_collate_fixmate_sort_markdup_bam.markdupstats \\\n",
      "                    --write-index \\\n",
      "                    \"$output_dir$out_name\"_collate_fixmate_sort_markdup.bam)\n",
      "\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"samtools collate > fixmate > sort > markdup COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat samtools_collate_fixmate_sort_markdup_HIMEM_230724_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1a6efc6-165c-4c03-9a09-4afd40e339da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 646005\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646002     nodes     bash   fmbuga  R      10:17      1 node26\n",
      "            646005     nodes samtools   fmbuga  R       0:05      1 node05\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646002     nodes     bash   fmbuga  R      10:27      1 node26\n",
      "            646005     nodes samtools   fmbuga  R       0:15      1 node05\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646002     nodes     bash   fmbuga  R      10:42      1 node26\n",
      "            646005     nodes samtools   fmbuga  R       0:30      1 node05\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646002     nodes     bash   fmbuga  R      11:12      1 node26\n",
      "            646005     nodes samtools   fmbuga  R       1:00      1 node05\n"
     ]
    }
   ],
   "source": [
    "# submit mark duplicates job to SLURM HPC\n",
    "!sbatch samtools_collate_fixmate_sort_markdup_230723_slurm.sh \\\n",
    "    ./01_raw_bam/ERR1831349.bam \\\n",
    "    ./02_bam_markdup/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0111c51b-6407-4bbb-8fb9-54639ecc1f91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 646007\n"
     ]
    }
   ],
   "source": [
    "# submit mark duplicates job to SLURM HPC\n",
    "!sbatch samtools_collate_fixmate_sort_markdup_HIMEM_230724_slurm.sh \\\n",
    "    ./01_raw_bam/ERR1831349.bam \\\n",
    "    ./02_bam_markdup/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "82c968d0-3c99-464d-a039-cb0725e422f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646002     nodes     bash   fmbuga  R      26:06      1 node26\n",
      "            646007     himem samtools   fmbuga  R       0:52      1 himem01\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646002     nodes     bash   fmbuga  R      26:16      1 node26\n",
      "            646007     himem samtools   fmbuga  R       1:02      1 himem01\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646002     nodes     bash   fmbuga  R      26:32      1 node26\n",
      "            646007     himem samtools   fmbuga  R       1:18      1 himem01\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646002     nodes     bash   fmbuga  R      27:02      1 node26\n",
      "            646007     himem samtools   fmbuga  R       1:48      1 himem01\n"
     ]
    }
   ],
   "source": [
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98c3faab-62c2-4e3f-9ec5-76d719b2ced4",
   "metadata": {},
   "source": [
    "- ERROR\n",
    "`[bam_sort_core] merging from 24 files and 12 in-memory blocks...\n",
    "[markdup] error: no ms score tag. Please run samtools fixmate on file first.\n",
    "[markdup] error: no ms score tag. Please run samtools fixmate on file first.\n",
    "`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ac96b88c-de30-4caa-99d5-38fe7ba61388",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 26 files and 26 in-memory blocks...\n",
      "[markdup] error: no ms score tag. Please run samtools fixmate on file first.\n",
      "[markdup] error: no ms score tag. Please run samtools fixmate on file first.\n",
      "samtools sort: failed writing to \"-\": Illegal seek\n"
     ]
    }
   ],
   "source": [
    "!samtools collate \\\n",
    "    -Ou \\\n",
    "    --threads 26 \\\n",
    "    ./01_raw_bam/ERR1831349.bam | \\\n",
    "        samtools fixmate \\\n",
    "            - \\\n",
    "            - \\\n",
    "            --threads 26 \\\n",
    "            -u | \\\n",
    "            samtools sort \\\n",
    "                - \\\n",
    "                --threads 26 \\\n",
    "                -u |\\\n",
    "                samtools markdup \\\n",
    "                    - \\\n",
    "                    --threads 26 \\\n",
    "                    input_collate_fixmate_sort_markdup.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7b2cd108-5014-45e5-9507-fccc68a2669e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: samtools sort [options...] [in.bam]\n",
      "Options:\n",
      "  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)\n",
      "  -u         Output uncompressed data (equivalent to -l 0)\n",
      "  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]\n",
      "  -M         Use minimiser for clustering unaligned/unplaced reads\n",
      "  -K INT     Kmer size to use for minimiser [20]\n",
      "  -n         Sort by read name (not compatible with samtools index command)\n",
      "  -t TAG     Sort by value of TAG. Uses position as secondary index (or read name if -n is set)\n",
      "  -o FILE    Write final output to FILE rather than standard output\n",
      "  -T PREFIX  Write temporary files to PREFIX.nnnn.bam\n",
      "  --no-PG    do not add a PG line\n",
      "      --input-fmt-option OPT[=VAL]\n",
      "               Specify a single input file format option in the form\n",
      "               of OPTION or OPTION=VALUE\n",
      "  -O, --output-fmt FORMAT[,OPT[=VAL]]...\n",
      "               Specify output format (SAM, BAM, CRAM)\n",
      "      --output-fmt-option OPT[=VAL]\n",
      "               Specify a single output file format option in the form\n",
      "               of OPTION or OPTION=VALUE\n",
      "      --reference FILE\n",
      "               Reference sequence FASTA FILE [null]\n",
      "  -@, --threads INT\n",
      "               Number of additional threads to use [0]\n",
      "      --write-index\n",
      "               Automatically index the output files [off]\n",
      "      --verbosity INT\n",
      "               Set level of verbosity\n"
     ]
    }
   ],
   "source": [
    "!samtools sort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6822a7f-841e-49be-9cdc-f16ec91ba558",
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools collate \\\n",
    "    -Ou \\\n",
    "    --threads 28 \\\n",
    "    ./01_raw_bam/ERR1831349.bam | \\\n",
    "        samtools fixmate \\\n",
    "            - \\\n",
    "            - \\\n",
    "            --threads 28 \\\n",
    "            -u | \\\n",
    "            samtools sort \\\n",
    "                - \\\n",
    "                --threads 28 \\\n",
    "                -u \\\n",
    "                input_collate_fixmate_sort.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a53be0cc-db64-46aa-b02c-8537cd2eeb1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=samtools_collate_fixmate_sort_markdup_230723\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_collate_fixmate_sort_markdup_230723_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_collate_fixmate_sort_markdup_230723_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=28\n",
      "#SBATCH --mem=0\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate minimap2_224                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "call=\"samtools --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "input_bam=$1\n",
      "output_dir=$2\n",
      "\n",
      "out_name=$(basename $input_bam .bam)\n",
      "\n",
      "call=$(samtools collate \\\n",
      "    -Ou \\\n",
      "    --threads 28 \\\n",
      "    $input_bam | \\\n",
      "        samtools fixmate \\\n",
      "            - \\\n",
      "            - \\\n",
      "            --threads 28 \\\n",
      "            -u | \\\n",
      "            samtools sort \\\n",
      "                - \\\n",
      "                --threads 28 \\\n",
      "                -u \\\n",
      "                -o \"$output_dir$out_name\"_collate_fixmate_sort.bam)\n",
      "\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"samtools collate > fixmate > sort COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat samtools_collate_fixmate_sort_230723_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "85da29d5-c3c3-460a-b787-82c3a5e0f312",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 646011\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R      52:10      1 node58\n",
      "            646011     nodes samtools   fmbuga  R       0:06      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R      52:20      1 node58\n",
      "            646011     nodes samtools   fmbuga  R       0:16      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R      52:35      1 node58\n",
      "            646011     nodes samtools   fmbuga  R       0:31      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R      53:05      1 node58\n",
      "            646011     nodes samtools   fmbuga  R       1:01      1 node25\n"
     ]
    }
   ],
   "source": [
    "# submit samtools_collate_fixmate_sort job to SLURM HPC\n",
    "!sbatch samtools_collate_fixmate_sort_230723_slurm.sh \\\n",
    "    ./01_raw_bam/ERR1831349.bam \\\n",
    "    ./02_bam_markdup/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "eea0b188-33c6-4e56-9a0b-3964159c73cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=samtools_markdup_230723\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_markdup_230723_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/samtools_markdup_230723_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=28\n",
      "#SBATCH --mem=0\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate minimap2_224                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "call=\"samtools --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "input_bam=$1\n",
      "output_dir=$2\n",
      "\n",
      "out_name=$(basename $input_bam .bam)\n",
      "\n",
      "call=$(samtools markdup \\\n",
      "                    $input_bam \\\n",
      "                    --threads 28 \\\n",
      "                    -sf \"$output_dir$out_name\"_markdup_bam.markdupstats \\\n",
      "                    --write-index \\\n",
      "                    \"$output_dir$out_name\"_markdup.bam)\n",
      "\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"samtools markdup COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat samtools_markdup_230723_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cfe57dfa-3537-4145-a2c7-5302ae57c100",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 646015\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R    1:16:25      1 node58\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R    1:16:35      1 node58\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R    1:16:50      1 node58\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R    1:17:21      1 node58\n"
     ]
    }
   ],
   "source": [
    "# submit mark duplicates job to SLURM HPC\n",
    "!sbatch samtools_markdup_230723_slurm.sh \\\n",
    "    ./02_bam_markdup/ERR1831349_collate_fixmate_sort.bam \\\n",
    "    ./02_bam_markdup/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c9cd0d7-edc9-4288-9e56-b45243c07012",
   "metadata": {},
   "source": [
    "```\n",
    "ERROR\n",
    "\n",
    "[markdup] error: no ms score tag. Please run samtools fixmate on file first.\n",
    "[markdup] error: no ms score tag. Please run samtools fixmate on file first.\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e6fb212c-b7b9-412e-a2cc-18d721ef1cda",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=gatk_markDuplicates_221003\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/gatk_markDuplicates_221003_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/gatk_markDuplicates_221003_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=28\n",
      "#SBATCH --mem=120G\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate gatk4                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "call=\"gatk MarkDuplicates --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "call=\"gatk MarkDuplicatesSpark --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "call=\"gatk MarkDuplicates --help\" # help\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "call=\"gatk MarkDuplicatesSpark --help\" # help\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "output_dir=\"./04_bam_dedup/\"            ### OUTPUT DIRECTORY \n",
      "echo \"output directory: $output_dir\"\n",
      "echo \"\"\n",
      "\n",
      "out_name=$(basename $1 .bam)\n",
      "echo $out_name\n",
      "\n",
      "echo $output_dir$out_name.bam\n",
      "\n",
      "# gatk MarkDuplicatesSpark\n",
      "\n",
      "call=$(gatk MarkDuplicatesSpark \\\n",
      "            -I $1 \\\n",
      "            -O \"$output_dir$out_name\"_dedup.bam \\\n",
      "            -M \"$output_dir$out_name\"_dedup_metrics.txt \\\n",
      "            --conf \"spark.local.dir=$output_dir\")        ### INPUT BAM\n",
      "                        ### Note if 'spark.executor.cores' is unset, Spark will use all available cores on the machine.\n",
      "                        ### https://gatk.broadinstitute.org/hc/en-us/articles/360037224932-MarkDuplicatesSpark\n",
      "                            ### \"This Spark tool requires a significant amount of disk operations. Run with both the input data and outputs on high throughput SSDs when possible. \n",
      "                            ### When pipelining this tool on Google Compute Engine instances, for best performance requisition machines with LOCAL SSDs.\"\n",
      "                            ### \"Furthermore, we recommend explicitly setting the Spark temp directory to an available SSD when \n",
      "                            ### running this in local mode by adding the argument --conf 'spark.local.dir=/PATH/TO/TEMP/DIR'.\"\n",
      "    \n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"marking duplicates COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat ~/tools/slurm_scripts/gatk_markDuplicates_221003_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "239a66b4-553e-4363-9f0d-6659cb74f591",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp ~/tools/slurm_scripts/gatk_markDuplicates_221003_slurm.sh \\\n",
    "    gatk_markDuplicates_230724_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "10a923b1-11e4-4c0e-8d6d-cc80c3a3099f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=gatk_markDuplicates_221003\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/gatk_markDuplicates_221003_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/gatk_markDuplicates_221003_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=28\n",
      "#SBATCH --mem=0\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "echo \"\"\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate gatk4                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "call=\"gatk MarkDuplicates --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "call=\"gatk MarkDuplicatesSpark --version\" # to get the version\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "call=\"gatk MarkDuplicates --help\" # help\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "call=\"gatk MarkDuplicatesSpark --help\" # help\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "input_bam=$1                            ### INPUT BAM\n",
      "output_dir=$2                           ### OUTPUT DIRECTORY \n",
      "\n",
      "\n",
      "echo \"output directory: $output_dir\"\n",
      "echo \"\"\n",
      "\n",
      "out_name=$(basename $input_bam .bam)\n",
      "echo $out_name\n",
      "\n",
      "echo $output_dir$out_name.bam\n",
      "\n",
      "# gatk MarkDuplicatesSpark\n",
      "\n",
      "call=$(gatk MarkDuplicatesSpark \\\n",
      "            -I $input_bam \\\n",
      "            -O \"$output_dir$out_name\"_markdup.bam \\\n",
      "            -M \"$output_dir$out_name\"_markdup_metrics.txt \\\n",
      "            --conf \"spark.local.dir=$output_dir\")        ### INPUT BAM\n",
      "                        ### Note if 'spark.executor.cores' is unset, Spark will use all available cores on the machine.\n",
      "                        ### https://gatk.broadinstitute.org/hc/en-us/articles/360037224932-MarkDuplicatesSpark\n",
      "                            ### \"This Spark tool requires a significant amount of disk operations. Run with both the input data and outputs on high throughput SSDs when possible. \n",
      "                            ### When pipelining this tool on Google Compute Engine instances, for best performance requisition machines with LOCAL SSDs.\"\n",
      "                            ### \"Furthermore, we recommend explicitly setting the Spark temp directory to an available SSD when \n",
      "                            ### running this in local mode by adding the argument --conf 'spark.local.dir=/PATH/TO/TEMP/DIR'.\"\n",
      "    \n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"marking duplicates COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat gatk_markDuplicates_230724_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23133fa4-b52c-4ddd-a0ec-2a1af4c24067",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 646016\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R    1:31:36      1 node58\n",
      "            646016     nodes gatk_mar   fmbuga  R       0:05      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R    1:31:46      1 node58\n",
      "            646016     nodes gatk_mar   fmbuga  R       0:15      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R    1:32:02      1 node58\n",
      "            646016     nodes gatk_mar   fmbuga  R       0:31      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646009     nodes     bash   fmbuga  R    1:32:32      1 node58\n",
      "            646016     nodes gatk_mar   fmbuga  R       1:01      1 node25\n"
     ]
    }
   ],
   "source": [
    "# submit mark duplicates job to SLURM HPC\n",
    "!sbatch gatk_markDuplicates_230724_slurm.sh \\\n",
    "    ./01_raw_bam/ERR1831349.bam \\\n",
    "    ./02_bam_markdup/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a978331c-2c85-4567-a250-12674d5af00b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93603695 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "0 + 0 secondary\n",
      "46995 + 0 supplementary\n",
      "34258723 + 0 duplicates\n",
      "93403023 + 0 mapped (99.79% : N/A)\n",
      "93556700 + 0 paired in sequencing\n",
      "46778350 + 0 read1\n",
      "46778350 + 0 read2\n",
      "92600884 + 0 properly paired (98.98% : N/A)\n",
      "93279060 + 0 with itself and mate mapped\n",
      "76968 + 0 singletons (0.08% : N/A)\n",
      "593850 + 0 with mate mapped to a different chr\n",
      "517790 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "!samtools flagstat \\\n",
    "    -@ 28 \\\n",
    "    ./02_bam_markdup/ERR1831349_markdup.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aafdb0c-76d5-4430-8bc0-f556ff0662a4",
   "metadata": {},
   "source": [
    "## Variant Calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5fe638bc-87df-4853-b84a-b580b58db2d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make folder to store vcf\n",
    "!mkdir 03_vcf_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "91a7afeb-ac9d-4112-acb7-2b9b2735788e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=bcftools_mpileup_call_ONT_P099_221129\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/bcftools_mpileup_call_ONT_P099_221129_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/bcftools_mpileup_call_ONT_P099_221129_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=1\n",
      "#SBATCH --mem-per-cpu=120G\n",
      "\n",
      "\n",
      "# \"...multithreading ...is currently used only for the compression of the output stream, only when --output-type is b or z\"\n",
      "# https://samtools.github.io/bcftools/bcftools.html\n",
      "# \"Use the -Ou option when piping between bcftools subcommands to speed up performance by removing unnecessary compression/decompression and VCF←→BCF conversion\"\n",
      "# for ONT https://github.com/samtools/bcftools/issues/1666\n",
      "\n",
      "# \"For ONT it may be beneficial to also run bcftools call with a higher -P, eg -P0.01 or -P 0.1\"\n",
      "\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate minimap2_224                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "call=\"bcftools --version\"\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "input_bam=$1\n",
      "filename=$(basename $1 .bam)\n",
      "output_dir=$2\n",
      "\n",
      "call=$(bcftools mpileup \\\n",
      "    -Ou \\\n",
      "    --config ont \\\n",
      "    -f ~/tools/hg38/hg38.fa \\\n",
      "    $input_bam |\\\n",
      "    bcftools call \\\n",
      "        --ploidy GRCh38 \\\n",
      "        -m -v \\\n",
      "        -P 0.99 \\\n",
      "        -o \"$output_dir$filename\"_bcftools_P_099.vcf) # INPUT BAM\n",
      "    \n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"bcftools mpileup call COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat ~/tools/slurm_scripts/bcftools_mpileup_call_ONT_P099_221128_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "81c7c676-759e-467d-82ec-f1ca4f9cf85e",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cp ~/tools/slurm_scripts/bcftools_mpileup_call_ONT_P099_221128_slurm.sh \\\n",
    "    bcftools_mpileup_call_P099_230724_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f6683b9-064f-4bd8-88d7-1b53e44e52f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#\n",
      "#SBATCH --job-name=bcftools_mpileup_call_ONT_P099_221129\n",
      "#SBATCH --output=/home/fmbuga/tools/slurm_scripts/slurm_out_err/bcftools_mpileup_call_ONT_P099_221129_%j.out\n",
      "#SBATCH --error=/home/fmbuga/tools/slurm_scripts/slurm_out_err/bcftools_mpileup_call_ONT_P099_221129_%j.err\n",
      "#\n",
      "#SBATCH --ntasks=1\n",
      "#SBATCH --cpus-per-task=1\n",
      "#SBATCH --mem=0\n",
      "\n",
      "\n",
      "# \"...multithreading ...is currently used only for the compression of the output stream, only when --output-type is b or z\"\n",
      "# https://samtools.github.io/bcftools/bcftools.html\n",
      "# \"Use the -Ou option when piping between bcftools subcommands to speed up performance by removing unnecessary compression/decompression and VCF←→BCF conversion\"\n",
      "# for ONT https://github.com/samtools/bcftools/issues/1666\n",
      "\n",
      "# \"For ONT it may be beneficial to also run bcftools call with a higher -P, eg -P0.01 or -P 0.1\"\n",
      "\n",
      "\n",
      "datenow=$(date)\n",
      "echo $datenow\n",
      "srun hostname\n",
      "echo \"\"\n",
      "\n",
      "start=$(date +%s)\n",
      "echo \"start time: $start\"\n",
      "echo \"hostname: $HOSTNAME\"\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "##################\n",
      "\n",
      "eval \"$(conda shell.bash hook)\"\n",
      "conda activate minimap2_224                      ### CONDA ENVIRONMENT NAME\n",
      "conda info\n",
      "echo \"\"\n",
      "\n",
      "\n",
      "call=\"bcftools --version\"\n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "input_bam=$1\n",
      "output_dir=$2\n",
      "\n",
      "filename=$(basename $1 .bam)\n",
      "\n",
      "call=$(bcftools mpileup \\\n",
      "    -Ou \\\n",
      "    -f ~/tools/hg38/hg38.fa \\\n",
      "    $input_bam |\\\n",
      "    bcftools call \\\n",
      "        --ploidy GRCh38 \\\n",
      "        -m -v \\\n",
      "        -P 0.99 \\\n",
      "        -o \"$output_dir$filename\"_bcftools_P_099.vcf) # INPUT BAM\n",
      "    \n",
      "echo $call\n",
      "eval $call\n",
      "echo \"\"\n",
      "\n",
      "echo \"bcftools mpileup call COMPLETE.\"\n",
      "echo \"\"\n",
      "    \n",
      "##################\n",
      "\n",
      "\n",
      "echo \"\"\n",
      "end=$(date +%s)\n",
      "echo \"end time: $end\"\n",
      "runtime_s=$(echo $(( end - start )))\n",
      "echo \"total run time(s): $runtime_s\"\n",
      "sec_per_min=60\n",
      "sec_per_hr=3600\n",
      "runtime_m=$(echo \"scale=2; $runtime_s / $sec_per_min;\" | bc)\n",
      "echo \"total run time(m): $runtime_m\"\n",
      "runtime_h=$(echo \"scale=2; $runtime_s / $sec_per_hr;\" | bc)\n",
      "echo \"total run time(h): $runtime_h\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cat bcftools_mpileup_call_P099_230724_slurm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4b0f1eb-8fc4-4d43-8306-639f65b42576",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 646019\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646017     nodes     bash   fmbuga  R      17:31      1 node58\n",
      "            646019     nodes bcftools   fmbuga  R       0:04      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646017     nodes     bash   fmbuga  R      17:42      1 node58\n",
      "            646019     nodes bcftools   fmbuga  R       0:15      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646017     nodes     bash   fmbuga  R      17:57      1 node58\n",
      "            646019     nodes bcftools   fmbuga  R       0:30      1 node25\n",
      "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
      "            646017     nodes     bash   fmbuga  R      18:27      1 node58\n",
      "            646019     nodes bcftools   fmbuga  R       1:00      1 node25\n"
     ]
    }
   ],
   "source": [
    "# submit bcftools mpileup call variants job to SLURM HPC\n",
    "!sbatch bcftools_mpileup_call_P099_230724_slurm.sh \\\n",
    "    ./02_bam_markdup/ERR1831349_markdup.bam \\\n",
    "    ./03_vcf_raw/\n",
    "\n",
    "!sleep 5\n",
    "!squeue -u fmbuga\n",
    "!sleep 10\n",
    "!squeue -u fmbuga\n",
    "!sleep 15\n",
    "!squeue -u fmbuga\n",
    "!sleep 30\n",
    "!squeue -u fmbuga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bdb7924-c3b0-43cd-bb25-12eea495fbe4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
